numpyro/mcmc.py

```diff
@@ -428,3 +441,303 @@ def single_chain_mcmc(rng, init_params):
+            with the input type to `potential_fn`.
+    :param model_args: Arguments provided to the model.
+    :param model_kwargs: Keyword arguments provided to the model.
+    """
+    @abstractmethod
+    def init(self, rng, num_warmup, init_params, model_args, model_kwargs):
+        raise NotImplementedError
+
+    @abstractmethod
+    def sample(self, state):
+        """
+        Given the current `state`, return the next `state` using the given
+        transition kernel.
+
+        :param state: Arbitrary data structure representing the state for the
+            kernel. For HMC, this is given by :data:`~numpyro.mcmc.HMCState`.
+        :return: Next `state`.
+        """
+        raise NotImplementedError
+
+
+class HMC(MCMCKernel):
+    """
+    Hamiltonian Monte Carlo inference, using fixed number of steps with
+    step size and mass matrix adaptation.
+
+    **References:**
+
+    1. *MCMC Using Hamiltonian Dynamics*,
+       Radford M. Neal
+
+    :param model: Python callable containing Pyro :mod:`~numpyro.primitives`.
+        If model is provided, `potential_fn` will be inferred using the model.
+    :param potential_fn: Python callable that computes the potential energy
+        given input parameters. The input parameters to `potential_fn` can be
+        any python collection type, provided that `init_params` argument to
+        `init_kernel` has the same type.
+    :param kinetic_fn: Python callable that returns the kinetic energy given
+        inverse mass matrix and momentum. If not provided, the default is
+        euclidean kinetic energy.
+    :param float step_size: Determines the size of a single step taken by the
+        verlet integrator while computing the trajectory using Hamiltonian
+        dynamics. If not specified, it will be set to 1.
+    :param bool adapt_step_size: A flag to decide if we want to adapt step_size
+        during warm-up phase using Dual Averaging scheme.
+    :param bool adapt_mass_matrix: A flag to decide if we want to adapt mass
+        matrix during warm-up phase using Welford scheme.
+    :param bool dense_mass:  A flag to decide if mass matrix is dense or
+        diagonal (default when ``dense_mass=False``)
+    :param float target_accept_prob: Target acceptance probability for step size
+        adaptation using Dual Averaging. Increasing this value will lead to a smaller
+        step size, hence the sampling will be slower but more robust. Default to 0.8.
+    :param float trajectory_length: Length of a MCMC trajectory for HMC. Default
+        value is :math:`2\\pi`.
+    """
+    def __init__(self,
+                 model=None,
+                 potential_fn=None,
+                 kinetic_fn=None,
+                 step_size=1.0,
+                 adapt_step_size=True,
+                 adapt_mass_matrix=True,
+                 dense_mass=False,
+                 target_accept_prob=0.8,
+                 trajectory_length=2*math.pi):
```

Comment:
could you also expose `max_tree_depth`?

Reply:
That's exposed in `NUTS`, which is basically subclassing off of this. I wanted to separate out the kernels so that we don't end up with bugs like `sampler='hmc', algo='nuts'`.